pkg/reconciler/service/resources/rolloutorchestrator.go

```diff
@@ -184,14 +227,42 @@ func NewInitialFinalTargetRev(initialRevisionStatus, ultimateRevisionTarget []v1
 }
 
 // UpdateInitialFinalTargetRev updates InitialRevisions, TargetRevisions and StageTargetRevisions for RolloutOrchestrator.
-func UpdateInitialFinalTargetRev(ultimateRevisionTarget []v1.TargetRevision, ro *v1.RolloutOrchestrator) {
+func UpdateInitialFinalTargetRev(ultimateRevisionTarget []v1.TargetRevision, ro *v1.RolloutOrchestrator,
+	route *servingv1.Route, deploymentLister appsv1listers.DeploymentLister) {
 	if !trafficEqual(ro.Spec.TargetRevisions, ultimateRevisionTarget) {
 		// If ultimateRevisionTarget is not equal to the TargetRevisions in the spec, it means the user updated the ksvc,
 		// leading to the new rollout, and the RolloutOrchestrator will start a new rollout, so we need to update
 		// the InitialRevisions, TargetRevisions and StageTargetRevisions.
 		if len(ro.Status.StageRevisionStatus) != 0 {
 			// Set the current StageRevisionStatus in status to the InitialRevisions.
-			ro.Spec.InitialRevisions = append([]v1.TargetRevision{}, ro.Status.StageRevisionStatus...)
+			if ro.IsStageReady() || ro.IsReady() {
+				// If we reach the end of the rollout or the stage rollout, set the Spec.InitialRevisions
+				// to Status.StageRevisionStatus directly.
+				ro.Spec.InitialRevisions = append([]v1.TargetRevision{}, ro.Status.StageRevisionStatus...)
+			} else {
+				if route == nil || len(route.Status.Traffic) == 0 || len(ro.Spec.StageTargetRevisions) == 0 {
+					// If route.Status.Traffic is empty, no revision is assigned to any traffic.
+					ro.Spec.InitialRevisions = nil
+				} else {
+					depName := ""
+					for _, rev := range ro.Spec.StageTargetRevisions {
+						if rev.IsRevScalingUp() {
+							// Locate the index of the revision scaling up
+							depName = fmt.Sprintf("%s-deployment", rev.RevisionName)
```

Comment:
This can be brittle as deployment name can be truncated, see cases like
revision name: `asr-whisper-finetuned-earnings-medium-predictor-00016`
deployment name: `asr-whisper-finetune07bd632c4ca7e99a731fb2010f241212-deployment`

Reply:
you can get the deployment using the label selector with revision name